src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs

```diff
@@ -60,28 +82,88 @@ internal ulong BucketSize
                 }
             }
 
-            // Relative cost of the bucket
-            // roughly proportional to the number of compares it needs in the success case.
-            internal int BucketCost
+            // if a bucket could be split into two degenerate ones
+            // specifies a label index where that split would occur
+            // -1 indicates that the bucket cannot be split into degenerate ones
+            //  0 indicates that the bucket is already degenerate
+            //
+            //  Degenerate buckets here are buckets with contiguous range of constants
+            //  leading to the same label. Like:
+            //
+            //      case 0:
+            //      case 1:
+            //      case 2:
+            //      case 3:
+            //           DoOneThing();
+            //           break;               
+            //
+            //      case 4:
+            //      case 5:
+            //      case 6:
+            //      case 7:
+            //           DoAnotherThing();
+            //           break;               
+            // 
+            // MOTIVATION:
+            // Table switch will perform two branches (range check and the actual computed jump).
+            // Computed jump costs the same or more expensive than a regular conditional branch.
+            // So compbined cost would be 2+ conditional branches.
+            // Based on benchmarks it is actually closer to 3
+            //
+            // On the other hand, handling a degenerate bucket requires only 1 compare (range check) 
+            // so crumbling one bucket into two degenerate ones, when possible, is advantageous.
+            //
+            // One relatively common case that may result in degenerate switches is in async/iterator 
+            // state machines when control flow must use cascading dispatch into try blocks
+            internal int DegenerateBucketSplit
             {
                 get
                 {
                     if (_startLabelIndex == _endLabelIndex)
```

Comment:
It would be the same as checking for buckets of size 1 as only those are known to be degenerate right after construction.
I guess it would be a bit cheaper to check for that

---
In reply to: [93715584](https://github.com/dotnet/roslyn/pull/16054#discussion_r93715584) [](ancestors = 93715584)